File: WEEKLY CONTEST 397/4.ts
import * as readline from 'readline';

interface TodoItem {
    name: string;
    completed: boolean;
}

class TodoList {
    private list: TodoItem[] = [];

    addToList(name: string) {
        let todoItem: TodoItem = {
            name: name,
            completed: false
        };
        this.list.push(todoItem);
    }

    printList() {
        this.list.forEach((item, index) => {
            let status = item.completed ? 'Done' : 'Not Done';
            console.log(`${index} ${item.name} - ${status}`);
        });
    }

    markDone(index: number) {
        if (index < this.list.length) {
            this.list[index].completed = true;
        }
    }
}

let todoList = new TodoList();
let rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function mainMenu() {
    console.log("1) Add to list");
    console.log("2) Mark done");
    console.log("3) Show list");
    console.log("4) Quit");

    rl.question("Choose an option: ", function(choice) {
        switch(choice) {
            case '1':
                rl.question("Enter your task: ", function(task) {
                    todoList.addToList(task);
                    mainMenu();
                });
                break;
            case '2':
                rl.question("Enter task number to mark as done: ", function(taskNum) {
                    todoList.markDone(parseInt(taskNum));
                    mainMenu();
                });
                break;
            case '3':
                todoList.printList();
                mainMenu();
                break;
            case '4':
                rl.close();
                break;
            default:
                console.log("Please enter a valid choice");
                mainMenu();
                break;
        }
    });
}

mainMenu();

File: WEEKLY CONTEST 397/4.rs
//Find the Minimum Cost Array Permutation in Todo-ist
use std::io::{self, Write};
use std::collections::LinkedList;

struct TodoItem {
    name: String,
    completed: char,
}

impl TodoItem {
    fn new(name: String) -> TodoItem {
        return TodoItem {
            name: name,
            completed: ' ',
        };
    }
}

struct TodoList {
    list: LinkedList<TodoItem>,
}

impl TodoList {
    fn new() -> TodoList {
        return TodoList {
            list: LinkedList::new(),
        };
    }

    fn add_to_list(&mut self, name: String) {
        let todo_item = TodoItem::new(name);
        self.list.push_back(todo_item);
    }

    fn print(&self) {
        for (i, item) in self.list.iter().enumerate() {
            println!("{} [{}] - {}", i, item.completed, item.name);
        }
    }

    fn mark_done(&mut self, index: usize) {
        let mut i = 0;
        for item in self.list.iter_mut() {
            if i == index {
                item.completed = 'x';
                break;
            }
            i += 1;
        }
    }
}

fn main() {
    let mut todo_list = TodoList::new();

    loop {
        println!("1) Add to list");
        println!("2) Mark done");
        println!("3) Show list");
        println!("4) Quit");

        let mut choice = String::new();
        io::stdin().read_line(&mut choice).expect("Failed to read line");

        let choice = choice.trim().parse::<u32>().expect("Please enter a number");

        match choice {
            1 => {
                print!("Enter your task: ");
                io::stdout().flush().expect("Failed to flush stdout");

                let mut task = String::new();
                io::stdin().read_line(&mut task).expect("Failed to read line");

                todo_list.add_to_list(task);
            },
            2 => {
                print!("Enter task number to mark as done: ");
                io::stdout().flush().expect("Failed to flush stdout");

                let mut task_num = String::new();
                io::stdin().read_line(&mut task_num).expect("Failed to read line");

                let task_num = task_num.trim().parse::<usize>().expect("Please enter a number");

                todo_list.mark_done(task_num);
            },
            3 => {
                todo_list.print();
            },
            4 => {
                break;
            },
            _ => {
                println!("Please enter a valid choice");
            },
        }
    }
}
